packages/graph/hash_graph/lib/graph/src/datastore/error.rs

```diff
@@ -0,0 +1,126 @@
+use std::{any::type_name, fmt, marker::PhantomData};
+
+use error_stack::{provider::Demand, Context};
+
+use crate::types::BaseId;
+
+#[derive(Debug)]
+#[must_use]
+pub struct InsertionError<T>(T);
+
+impl<T> InsertionError<T> {
+    pub const fn new(t: T) -> Self {
+        Self(t)
+    }
+}
+
+impl<T> fmt::Display for InsertionError<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(fmt, "Could not insert `{}` into database", type_name::<T>())
+    }
+}
+
+impl<T: fmt::Debug + Send + Sync + 'static> Context for InsertionError<T> {
+    fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
+        demand.provide_ref(&self.0);
+    }
+}
+
+// TODO: Remove `Q` generic by adding a trait to `T`
+#[must_use]
+pub struct QueryError<Q, T>(Q, PhantomData<T>);
```

Comment:
This struct is a little complex, why do we need it? (As in why the `PhantomData` and such).

Is it because we don't have a trait bound at the moment for the Types?

Does Q need to be generic, it seems like it's always `VersionId` no?

Reply:
Can we not just attach these things and not have this complicated code?